Revamped video URL parsing for PHP and JS #2955

bastianallgeier opened the pull request:
## Describe the PR

The parser for video URLs now uses the URI class in PHP and the URL parser in JS. They are actually pretty close, which is very cool. The code is almost identical that way and we can simply adapt it on both ends if needed. Tests have also been ported to JS to test the same URL patterns on both sides. 

## Related issues

- Fixes #2955

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Fixed code style issues with CS fixer and `composer fix`
- [ ] Added in-code documentation (if needed)

bastianallgeier committed:
Revamped video URL parsing for PHP and JS #2955

bastianallgeier committed:
Remove empty line

lukasbestle commented on panel/src/helpers/embed.js:
```diff
@@ -0,0 +1,131 @@
+    let query = uri.searchParams;
+    let src = null;
+
+    switch (path.join("/")) {
```
When reading the code, I was confused what happens with the query params in the resulting URL, because I didn't yet know they were appended at the end.

Suggestion: A quick comment above this that explains what happens:

```suggestion

    // build the correct base URL for the embed,
    // the query params are appended below
    let src = null;
    switch (path.join("/")) {
```

lukasbestle commented on panel/src/helpers/embed.js:
```diff
@@ -0,0 +1,131 @@
+    const queryString = query.toString();
+
+    if (queryString.length) {
+      src += "?" + queryString.replace("&", "&amp;");
```
For cleanliness I think this helper shouldn't escape the string as the returned URL is not a valid URL anymore. Does anything speak against escaping when the URL is used in the component?

bastianallgeier replied:
The only reason for escaping it here is to keep the results identical with the ones in PHP.

bastianallgeier replied:
I just fixed it. I think we should also stop encoding ampersands in PHP. I'm not sure why we are doing it anyway. But this should be done in a different PR. I don't know how deep that problem will go.

lukasbestle commented on src/Toolkit/Html.php:
```diff
@@ -446,128 +447,110 @@ public static function video(string $url, array $options = [], array $attr = [])
      * @return string The generated HTML
      */
-    public static function youtube(string $url, array $options = [], array $attr = []): string
+    public static function youtube(string $url, array $options = [], array $attrs = []): string
```
Renaming the param would be a breaking-change because of PHP 8's named arguments. Also it's only there for the YouTube method, not for Vimeo. Is the change really necessary?

bastianallgeier replied:
Ah, I still have to get used to the PHP 8 stuff. No, it's not necessary. I will revert it.

lukasbestle commented on src/Toolkit/Html.php:
```diff
@@ -446,128 +447,110 @@ public static function video(string $url, array $options = [], array $attr = [])
+                break;
+        }
+
+        if (!preg_match('!^[0-9]*$!', $id)) {
```
```suggestion
        if (preg_match('!^[0-9]*$!', $id) !== 1) {
```

lukasbestle commented on src/Toolkit/Html.php:
```diff
@@ -446,128 +447,110 @@ public static function video(string $url, array $options = [], array $attr = [])
+        $src    = null;
+
+        $isYoutubeId = function (string $id): bool {
+            return preg_match('!^[a-zA-Z0-9_-]+$!', $id);
```
`preg_match()` does not return a bool, but `1|0|false`. So Psalm will complain about this on higher levels. Fix:

```suggestion
            return preg_match('!^[a-zA-Z0-9_-]+$!', $id) === 1;
```

lukasbestle commented on src/Toolkit/Html.php:
```diff
@@ -446,128 +447,110 @@ public static function video(string $url, array $options = [], array $attr = [])
+
+                    $query->start = $query->t;
+
+                    unset($query->v, $query->t);
```
```suggestion
                    unset($query->v, $query->t);
```

lukasbestle commented on src/Toolkit/Html.php:
```diff
@@ -446,128 +447,110 @@ public static function video(string $url, array $options = [], array $attr = [])
 
-        $url = 'https://' . $domain . '/' . $uri . $id . $query;
+        // allow fullscreen mode by default
+        $attrs = array_merge(['allowfullscreen' => true], $attrs);
```
This change (extracting the `array_merge()` into its own line) is not mirrored in the `vimeo()` method. In contrast, the `vimeo()` method already appends the query string. I think the implementation should be mostly consistent for better readability.

lukasbestle reviewed (commented):
It would be great to also support other video platforms/embed types in the future, see https://kirby.nolt.io/170. I'm torn if we should use this opportunity to already offer this feature. But in the end we can still support it later, there's nothing that really stops us from fixing it first and then improving it again later.

My comments on the YouTube JS method also apply to the Vimeo one that uses identical code.

bastianallgeier committed:
Better inline docs

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

bastianallgeier committed:
Improve preg_match check

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

bastianallgeier committed:
Unify the code for the vimeo and youtube methods

bastianallgeier committed:
Improve coding style

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

bastianallgeier committed:
No longer encode ampersands in URLs

bastianallgeier committed:
More reliable check for video IDs.

bastianallgeier committed:
Better unit tests and another fix in the vimeo parsers

bastianallgeier committed:
Video methods no longer throw exceptions

To be used in templates more easily and to be more close to the JS counterpart, video, youtube and vimeo methods no longer throw exceptions but return null if the video URL is invalid.

bastianallgeier committed:
More unit tests

afbora committed:
Adds unit tests for blocks #2955

bastianallgeier committed:
Handle local video files in Html::video()

distantnative reviewed (approved):
LGTM
